pkg/sentry/fsimpl/fuse/dev.go

```diff
@@ -46,55 +49,230 @@ type DeviceFD struct {
+
 }
 
 // Release implements vfs.FileDescriptionImpl.Release.
 func (fd *DeviceFD) Release() {}
 
 // PRead implements vfs.FileDescriptionImpl.PRead.
 func (fd *DeviceFD) PRead(ctx context.Context, dst usermem.IOSequence, offset int64, opts vfs.ReadOptions) (int64, error) {
+	// Operations on /dev/fuse don't make sense until a FUSE filesystem is mounted.
+	if !fd.mounted {
+		return 0, syserror.EPERM
+	}
+
 	return 0, syserror.ENOSYS
 }
 
 // Read implements vfs.FileDescriptionImpl.Read.
 func (fd *DeviceFD) Read(ctx context.Context, dst usermem.IOSequence, opts vfs.ReadOptions) (int64, error) {
-	return 0, syserror.ENOSYS
+	// Operations on /dev/fuse don't make sense until a FUSE filesystem is mounted.
+	if !fd.mounted {
+		return 0, syserror.EPERM
+	}
+
+	kernelTask := kernel.TaskFromContext(ctx)
+	if kernelTask == nil {
+		log.Warningf("fusefs.DeviceFD.Read: couldn't get kernel task from context")
+		return 0, syserror.EINVAL
+	}
+
+	// Wait for a request to be made available.
+	if err := kernelTask.Block(fd.waitCh); err != nil {
+		log.Warningf("fusefs.DeviceFD.Read: couldn't wait on request queue: %v", err)
+		return 0, syserror.EBUSY
+	}
+
+	fd.mu.Lock()
+	defer fd.mu.Unlock()
+	return fd.readLocked(ctx, dst, opts)
+}
+
+// readLocked
+func (fd *DeviceFD) readLocked(ctx context.Context, dst usermem.IOSequence, opts vfs.ReadOptions) (int64, error) {
+	if fd.queue.Empty() {
+		log.Warningf("fusefs.DeviceFD.Read: No requests to read but still signalled")
+		return 0, syserror.EAGAIN
+	}
+
+	req := fd.queue.Front()
+	if fd.readCursor >= req.hdr.Len {
+		// Cursor points past end of current request payload? Reset the cursor,
+		// remove the front request and try again.
+		fd.readCursor = 0
+		fd.queue.Remove(req)
+		return fd.readLocked(ctx, dst, opts)
+	}
+
+	n, err := dst.CopyOut(ctx, req.data[fd.readCursor:])
+	fd.readCursor += uint32(n)
+
+	if fd.readCursor >= req.hdr.Len {
+		// Fully done with this req, remove it from the queue.
+		fd.queue.Remove(req)
+		fd.readCursor = 0
+	}
+
+	return int64(n), err
 }
 
 // PWrite implements vfs.FileDescriptionImpl.PWrite.
 func (fd *DeviceFD) PWrite(ctx context.Context, src usermem.IOSequence, offset int64, opts vfs.WriteOptions) (int64, error) {
+	// Operations on /dev/fuse don't make sense until a FUSE filesystem is mounted.
+	if !fd.mounted {
+		return 0, syserror.EPERM
+	}
+
 	return 0, syserror.ENOSYS
 }
 
 // Write implements vfs.FileDescriptionImpl.Write.
 func (fd *DeviceFD) Write(ctx context.Context, src usermem.IOSequence, opts vfs.WriteOptions) (int64, error) {
-	return 0, syserror.ENOSYS
+	fd.mu.Lock()
+	defer fd.mu.Unlock()
+	return fd.writeLocked(ctx, src, opts)
 }
 
-// Seek implements vfs.FileDescriptionImpl.Seek.
-func (fd *DeviceFD) Seek(ctx context.Context, offset int64, whence int32) (int64, error) {
-	return 0, syserror.ENOSYS
+// writeLocked
+func (fd *DeviceFD) writeLocked(ctx context.Context, src usermem.IOSequence, opts vfs.WriteOptions) (int64, error) {
```

Comment:
Naming is a little bit confusing. At the first sight, it seems to me that this function handles locks

Reply:
This is a common pattern to explicitly state that the method is called while holding a lock. Added a comment to call this out too. But this pattern is used all over.